File: ios/Sendro/Core/HistoryStore.swift
//
//  HistoryStore.swift
//  Sendro
//
//  Local JSON transfer history, newest first, capped at 500 entries.
//

import Foundation
import Combine

struct HistoryEntry: Codable, Identifiable {
    let id: String              // unique per entry (UUID)
    let transferId: String
    let fileName: String
    let sizeBytes: Int64
    let senderName: String
    let dateMs: Int64
    /// completed | failed | rejected | cancelled
    let outcome: String
    let savedTo: String?        // photos | files | temp
    let errorMessage: String?
    /// nil (legacy / incoming) | "outgoing" (iPhone → PC upload, §7).
    /// Optional so entries persisted before this field existed still decode.
    let direction: String?
    /// Where the bytes ended up on this iPhone, when they are still here:
    /// the file name inside Documents/Sendro (NOT an absolute path — the app
    /// container path changes on every reinstall, so storing one would rot).
    /// nil for media that was imported into Photos and deleted locally.
    let localName: String?
    /// PHAsset localIdentifier of the imported asset, when it went to Photos.
    /// Lets the preview fetch a thumbnail / full image straight from the
    /// library instead of showing a broken preview for a deleted temp file.
    let photoAssetId: String?
}

final class HistoryStore: ObservableObject {

    static let maxEntries = 500

    @Published private(set) var entries: [HistoryEntry] = []

    init() {
        load()
    }

    func add(transferId: String,
             fileName: String,
             sizeBytes: Int64,
             senderName: String,
             outcome: String,
             savedTo: String? = nil,
             errorMessage: String? = nil,
             direction: String? = nil,
             localName: String? = nil,
             photoAssetId: String? = nil) {
        let entry = HistoryEntry(id: UUID().uuidString,
                                 transferId: transferId,
                                 fileName: fileName,
                                 sizeBytes: sizeBytes,
                                 senderName: senderName,
                                 dateMs: Int64(Date().timeIntervalSince1970 * 1000),
                                 outcome: outcome,
                                 savedTo: savedTo,
                                 errorMessage: errorMessage,
                                 direction: direction,
                                 localName: localName,
                                 photoAssetId: photoAssetId)
        entries.insert(entry, at: 0)
        if entries.count > Self.maxEntries {
            entries = Array(entries.prefix(Self.maxEntries))
        }
        save()
    }

    func clear() {
        entries = []
        save()
    }

    private func load() {
        guard let data = try? Data(contentsOf: AppPaths.historyURL) else { return }
        if let decoded = try? JSONDecoder().decode([HistoryEntry].self, from: data) {
            entries = decoded
        }
    }

    private func save() {
        guard let data = try? JSONEncoder().encode(entries) else { return }
        try? data.write(to: AppPaths.historyURL, options: .atomic)
    }
}

File: android/app/src/main/java/com/sendro/android/core/HistoryStore.kt
package com.sendro.android.core

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import java.io.File
import java.util.UUID

/**
 * One finished transfer, in either direction.
 *
 * NOTE (§11): text messages are NEVER written here. This file is the reason
 * that rule needs restating — the temptation to "just log the message" ends
 * with plaintext on disk.
 */
@Serializable
data class HistoryEntry(
    val id: String,
    val transferId: String,
    val fileName: String,
    val sizeBytes: Long,
    val senderName: String,
    val dateMs: Long,
    /** completed | failed | rejected | cancelled */
    val outcome: String,
    /** photos | files | temp */
    val savedTo: String? = null,
    val errorMessage: String? = null,
    /** null (incoming) | "outgoing" (phone -> PC upload, §7) */
    val direction: String? = null,
    /**
     * The file name inside `files/received` when the bytes are still here —
     * NOT an absolute path, because the app's data dir changes on reinstall.
     */
    val localName: String? = null,
    /**
     * MediaStore content URI of the published asset, when it went to the
     * gallery. Lets the Library preview pull the image back out of MediaStore
     * after the local temp copy was deleted.
     */
    val mediaUri: String? = null,
)

/** Local transfer history, newest first, capped at 500 entries. */
class HistoryStore(private val file: File) {

    private val _entries = MutableStateFlow(load())
    val entries: StateFlow<List<HistoryEntry>> = _entries.asStateFlow()

    @Synchronized
    fun add(
        transferId: String,
        fileName: String,
        sizeBytes: Long,
        senderName: String,
        outcome: String,
        savedTo: String? = null,
        errorMessage: String? = null,
        direction: String? = null,
        localName: String? = null,
        mediaUri: String? = null,
    ) {
        val entry = HistoryEntry(
            id = UUID.randomUUID().toString(),
            transferId = transferId,
            fileName = fileName,
            sizeBytes = sizeBytes,
            senderName = senderName,
            dateMs = System.currentTimeMillis(),
            outcome = outcome,
            savedTo = savedTo,
            errorMessage = errorMessage,
            direction = direction,
            localName = localName,
            mediaUri = mediaUri,
        )
        val next = (listOf(entry) + _entries.value).take(MAX_ENTRIES)
        _entries.value = next
        save(next)
    }

    @Synchronized
    fun clear() {
        _entries.value = emptyList()
        save(emptyList())
    }

    @Synchronized
    fun removeById(id: String) {
        val next = _entries.value.filterNot { it.id == id }
        _entries.value = next
        save(next)
    }

    private fun load(): List<HistoryEntry> {
        if (!file.isFile) return emptyList()
        return runCatching { SendroJson.decodeFromString<List<HistoryEntry>>(file.readText()) }
            .getOrDefault(emptyList())
    }

    private fun save(list: List<HistoryEntry>) {
        runCatching {
            file.parentFile?.mkdirs()
            // Atomic-ish: write beside, then rename, so a kill mid-write never
            // leaves a truncated history file.
            val temp = File(file.parentFile, file.name + ".tmp")
            temp.writeText(SendroJson.encodeToString(list))
            if (!temp.renameTo(file)) {
                file.writeText(temp.readText())
                temp.delete()
            }
        }
    }

    companion object {
        const val MAX_ENTRIES = 500
    }
}
